crates/matrix-sdk-base/src/store/mod.rs

```diff
@@ -139,12 +140,63 @@ pub(crate) struct Store {
     /// The current sync token that should be used for the next sync call.
     pub(super) sync_token: Arc<RwLock<Option<String>>>,
     /// All rooms the store knows about.
-    rooms: Arc<StdRwLock<BTreeMap<OwnedRoomId, Room>>>,
+    rooms: Arc<StdRwLock<Rooms>>,
     /// A lock to synchronize access to the store, such that data by the sync is
     /// never overwritten.
     sync_lock: Arc<Mutex<()>>,
 }
 
+#[derive(Default, Debug)]
+struct Rooms {
+    mapping: BTreeMap<OwnedRoomId, usize>,
+    rooms: ObservableVector<Room>,
+}
+
+impl Rooms {
+    fn insert(&mut self, room_id: OwnedRoomId, room: Room) -> usize {
+        match self.mapping.get(&room_id) {
+            Some(position) => {
+                self.rooms.set(*position, room);
+
+                *position
+            }
+            None => {
+                let position = self.rooms.len();
+
+                self.rooms.push_back(room);
+                self.mapping.insert(room_id, position);
+
+                position
+            }
+        }
+    }
+
+    fn get(&self, room_id: &RoomId) -> Option<&Room> {
+        self.mapping.get(room_id).and_then(|position| self.rooms.get(*position))
+    }
+
+    fn get_or_create<F>(&mut self, room_id: &RoomId, default: F) -> &Room
+    where
+        F: FnOnce() -> Room,
+    {
+        let position = match self.mapping.get(room_id) {
+            Some(position) => *position,
+            None => {
+                let room = default();
+                self.insert(room.room_id().to_owned(), room)
```

Comment:
We can do better than this: `self.insert` will re-do the `self.mapping.get()` lookup, which is unnecessary. Can you inline the `None` handling case of `insert` here instead directly, instead?

Reply:
Done.